.github/workflows/main.yml

```diff
@@ -97,3 +72,128 @@ jobs:
             -m "Build results of ${GITHUB_REF#refs/heads/*} kaitai-io/ci_targets@$GITHUB_SHA" \
             -b "$TARGET$SUBTARGET/$VARIETY" \
             -- --exclude=.git --exclude=.travis.yml tests/test_out
+
+  dockerized_test:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        tsv:
+          - target: cpp_stl
+            subtarget: _98
+            implementation: gcc4.8
+          - target: cpp_stl
+            subtarget: _11
+            implementation: gcc4.8
+          - target: cpp_stl
+            subtarget: _98
+            implementation: gcc11
+#          - target: cpp_stl
+#            subtarget: _11
+#            implementation: gcc11
+          - target: java
+            subtarget: ''
+            implementation: zulu7
+          - target: java
+            subtarget: ''
+            implementation: temurin8
+          - target: java
+            subtarget: ''
+            implementation: temurin11
+          - target: java
+            subtarget: ''
+            implementation: temurin17
+          - target: javascript
+            subtarget: ''
+            implementation: nodejs4
+          - target: javascript
+            subtarget: ''
+            implementation: nodejs8
+          - target: javascript
+            subtarget: ''
+            implementation: nodejs10
+          - target: javascript
+            subtarget: ''
+            implementation: nodejs12
+          - target: javascript
+            subtarget: ''
+            implementation: nodejs18
+          - target: python
+            subtarget: ''
+            implementation: 2.7
+          - target: python
+            subtarget: ''
+            implementation: 3.4
+          - target: python
+            subtarget: ''
+            implementation: 3.11
+          - target: ruby
+            subtarget: ''
+            implementation: 1.9
+          - target: ruby
+            subtarget: ''
+            implementation: 2.3
+          - target: ruby
+            subtarget: ''
+            implementation: 3.2
+    steps:
```

Comment:
```suggestion
    env:
      TARGET: ${{matrix.tsv.target}}
      SUBTARGET: ${{matrix.tsv.subtarget}}
      IMPL: ${{matrix.tsv.implementation}}
    steps:
```

```diff
       - name: docker pull
         run: |
-          docker pull ghcr.io/kaitai-io/kaitai-${{matrix.tsv.target}}-${{matrix.tsv.implementation}}-linux-x86_64
+          docker pull ghcr.io/kaitai-io/kaitai-"$TARGET"-"$IMPL"-linux-x86_64
```

```diff
       - name: download runtime
         run: |
           mkdir -p runtime
-          git clone https://github.com/kaitai-io/kaitai_struct_${{matrix.tsv.target}}_runtime runtime/${{matrix.tsv.target}}
+          git clone https://github.com/kaitai-io/kaitai_struct_"$TARGET"_runtime runtime/"$TARGET"
```

```diff
         run: |
           cd tests
-          ./docker-ci -t "${{matrix.tsv.target}}" -u "${{matrix.tsv.subtarget}}" -i "${{matrix.tsv.implementation}}"
+          ./docker-ci -t "$TARGET" -u "$SUBTARGET" -i "$IMPL"
       # - name: publish as artifact
       #   uses: actions/upload-artifact@v3
       #   with:
       #     name: test-${{matrix.tsv.target}}${{matrix.tsv.subtarget}}-${{matrix.tsv.implementation}}
       #     path: tests/test_out
       - name: publish
         env:
           BOT_SSH_KEY: ${{secrets.BOT_SSH_KEY}}
-          TARGET: ${{matrix.tsv.target}}
-          SUBTARGET: ${{matrix.tsv.subtarget}}
-          VARIETY: ${{matrix.tsv.implementation}}
         run: |
           ./push_artifacts/git_config_kaitai_bot
           ./push_artifacts/publish \
             -o kaitai-io \
             -r ci_artifacts \
             -m "Build results of ${GITHUB_REF#refs/heads/*} kaitai-io/ci_targets@$GITHUB_SHA" \
-            -b "$TARGET$SUBTARGET/$VARIETY" \
+             -b "${TARGET}${SUBTARGET}"/"$IMPL" \
             -- --exclude=.git --exclude=.travis.yml tests/test_out
```

Reply:
@GreyCat Also check out https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections.

Reply:
Thanks! On the security aspects of it, I believe I've set up this repo to have any PR requests build to be approved by Kaitai team members. So, if we see that somebody wants to do something fishy with anything in the pipeline in relation to BOT_SSH_KEY, we really don't want to approve launching it.